### ISSUE TITLE
Does Doxygen have a Typescript Parser ?

### ISSUE THREAD

**Issue by florentinGo** [doxygen/doxygen#9904]
Hello everyone,

I have a project that combines C# source code (back end) and Typescript source code (front end). Doxygen works perfectly for C#, but I am wondering if there is a solution to handle the Typescript source code.

I have tried using the Javascript syntax to parse the Typescript, but the behavior seems to be inconsistent. Sometimes it finds the method and sometimes it does not.

Should I use a filter or is there an existing parser that can handle Typescript source code?

Thank you for your hard work.

**Comment by albert-github** [comment]
from https://en.wikipedia.org/wiki/TypeScript I see:
> TypeScript is a [free and open source](https://en.wikipedia.org/wiki/Free_and_open_source) [high-level](https://en.wikipedia.org/wiki/High-level_programming_language) [programming language](https://en.wikipedia.org/wiki/Programming_language) developed and maintained by [Microsoft](https://en.wikipedia.org/wiki/Microsoft). It is a strict syntactical [superset](https://en.wikipedia.org/wiki/Superset) of [JavaScript](https://en.wikipedia.org/wiki/JavaScript) and adds optional [static typing](https://en.wikipedia.org/wiki/Static_typing) to the language.

There is no support for TypeScript in doxygen, though there is some, old, rudimentary support for JavaScript, for mapping the extension see the setting `EXTENSION_MAPPING` (https://www.doxygen.nl/manual/config.html#cfg_extension_mapping) where your extension is mapped to `javascript`.
Most likely resorting to a filter would be the best choice.

**Comment by Hi-Angel** [comment]
I'm wondering what info the "needinfo" tag refers to. 

I think this could be labeled as a feature-request in case someone wants to add support for TypeScript.

**Comment by yjfvictor** [comment]
No. It doesn't support TypeScript at all. TypeScrpt code is wrongly parsed. All the TypeScript variables are wrongly shown with lots of pads such as `__pad0__`, even if I have tried to use `@var` to correct the variable name.

![Doxygen parsing errors](https://github.com/user-attachments/assets/a228a623-7da8-483c-b22a-c419d90f4514)

![Doxygen parsing errors](https://github.com/user-attachments/assets/0fc81590-c826-404a-be3c-acecdadf2ecd)

The source code is like:

![source code](https://github.com/user-attachments/assets/14a712b7-f739-4008-be38-5ac7dca4f1aa)

**Comment by albert-github** [comment]
@yjfvictor 

- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file!) that allows us to reproduce the problem?
   Please don't add external links as they might not be persistent (also references to GitHub repositories are considered non persistent).
- Please also specify the full doxygen version used (`doxygen -v`).

**Comment by yjfvictor** [comment]
@albert-github 

Here is a small, self-contained example as you like: [doxygen-test.zip](https://github.com/user-attachments/files/25389196/doxygen-test.zip). You can download this zip file.


The Doxygen version number is 1.9.8 on Linux and 1.16.1 on Windows.

Both of them wrongly make documents with lots of pads in the interface `Person`, such as `__pad0__`, `__pad1__`, `__pad2__`, `__pad3__`.
![documents with lots of pads](https://github.com/user-attachments/assets/ec32191b-3bf4-4a3d-b8ab-bbf28189c7d2)

Moreover, it cannot show the member variables of class `Calculator`.

![class member not showing](https://github.com/user-attachments/assets/db02d85a-a2e4-43b3-b8b0-14c516f003ae)

**Comment by yjfvictor** [comment]
There exist two different styles of variable declaration:

1. **prefix typing**, e.g. `int num;` in C, C++, C#, Java, and `integer :: num` in Fortran.
2. **postfix typing**, e.g. `var num int` in Golang, `var num: number;` in TypeScript, `num: int` in Python 3.5+ and Pascal, `Dim num As Integer` in freeBasic.

I wonder whether Doxygen's author is aware of these differences. According to my tests on TypeScript, Doxygen seems to wrongly regard TypeScript as a prefix-typing language, while it is actually a postfix-typing language.

My TypeScript source code contains `readonly createdAt: string;`, so the variable is `createdAt`, and the type is `string`, but Doxygen wrongly regards `readonly` as the type, and therefore is wrongly displayed as <code>readonly **createdAt**</code>.

![TypeScript source code](https://github.com/user-attachments/assets/eff26b39-622a-4b79-ad92-7dc9f0869d7a)

![Doxygen-generated documents](https://github.com/user-attachments/assets/fc6a6e30-2dfd-4664-87ca-f4c7acf223f9)
